### ISSUE TITLE
macOS DNS & Firewall bug

### ISSUE THREAD

**Issue by raggi** [tailscale/tailscale#13511]
There are few concrete facts known so far.

Symptoms:

* On macOS prior to Sequoia: an MDM profile installed firewall configuration of "block all incoming connections" transiently blocks UDP sessions - sometimes this is very sticky on a machine, other times it appears to occur only after some arbitrary amount of runtime / after some uknown trigger. On this system version we observe magicsock reporting sendto(2) returning `EPERM`. In one reproduction case we've witnessed directly on this macOS version DERP connections keep working ok through this, so the overall scenario is just a substantial performance degradation.
* On macOS Sequoia:
  * We have seen two instances where a system is running perfectly fine, then at some arbitrary point the first noticeable symptom is DNS starts to lock up, followed by many other connections. `curl http://example.org` will for example hang indefinitely in this state. A second reproduction of this variant occurred on a system with SIP disabled. After starting dtruss of the curl process, console.app, finder.app, and the terminal program hosting the dtruss invocation all locked up hard, beachballing. New SSH sessions were accept'd, but did not spawn sessions. We wanted to gather more data from this state, but ran out of ways to interact with the system.
  * We have seen two independent instances on different machines of DNS becoming "wonky", but other sockets not failing. On both of these instances dig against quad-100 still works fine, but DNS resolution via the system APIs hangs. One of the two systems observed eventaul resolution timeouts, the other progressively locked up entirely (I had to go bio and get a drink after hours of this, and when I came back logging in just span indefinitely - replugging power lead to just the battery display left and no way to interact with the system - hard kernel lock up again it seems, like the above).

Trigger conditions:

As yet we don't know how to actively trigger this/these bug(s). We do know that they are substantially more likely to trigger if the firewall is configured with "block all incoming traffic" - we've only seen repro's on systems with this setting set. If tailscale is stopped while the symptoms are mild (some dns failures), then the symptoms entirely subside until tailscale is enabled again.

Internal bug: tailscale/corp#23225

**Comment by florian-obradovic** [comment]
Seems like a general BUG in macOS Sequoia & Firewall.

We had general DNS issues after a Sonoma > Sequoia upgrade (no Tailscale involved).
For Example: Terminal.app was blocked so no ping, nslookup, etc. was working. Other apps also didn't work but a few did like Chrome, Safari, etc.

```
/usr/libexec/ApplicationFirewall/socketfilterfw --listapps
Total number of apps = 6 
1 : /System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal 
             (Block incoming connections)
```
I wasn't able to --remove the Terminal .app but was able to fix it via:
`/usr/libexec/ApplicationFirewall/socketfilterfw --add /System/Applications/Utilities/Terminal.app/Contents/MacOS/Terminal `

It was really strange. 
The biggest issue here was that the firewall was set via Intune MDM Configuration profile and the Intune Company Portal didn't sync new policies due to no working DNS :)
So I had to find a way to remove all profiles to be able to manually configure the firewall which was a real pain with a user sitting remotely.

You basically have to disable SIP (System Integrity Protection) by booting into recovery mode, remove all profiles, re-enroll the Mac in your MDM.

Shut down the machine normally and then do the following depending on what kind of Mac you have:

- Intel Mac, hold down the “Command” and “R” buttons on the keyboard while turning on the Mac and keep holding them until the Apple logo and progress bar shows

- Apple Silicon/M1 Mac, press and hold the power button. Wait for "keep holding to get boot Options"
- Click "Options" to boot into Recovery Mode

In Recovery Mode, open the Terminal via the Utilities menu
![CleanShot 2024-09-19 at 09 40 49@2x](https://github.com/user-attachments/assets/18af8047-1efd-4d00-b4f1-1401847b714f)
  
`csrutil disable`

Now reboot, SIP is now disabled.

Login and follow these steps:

![CleanShot 2024-09-19 at 09 43 09@2x](https://github.com/user-attachments/assets/c7e4a119-eb5c-49b0-8d9b-91a6eb3967ca)

```
sudo rm -rf /var/db/ConfigurationProfiles/Store/*
sudo profiles -D
```

Now check System Settings > General > Device Management / Profiles: It should be empty now.
Reboot and check profiles again. If there are some left, you should be able to delete theme.

Now I was able to configure the firewall manually and everything was working again
![CleanShot 2024-09-19 at 09 45 58@2x](https://github.com/user-attachments/assets/ae4f130f-d01a-4255-9734-15e30089f9f1)


Boot back into Recovery Mode and enable SIP again:
`csrutil enable`

Start Intune Company Portal, remove the device and enrol it again:
![CleanShot 2024-09-19 at 09 47 32@2x](https://github.com/user-attachments/assets/32e52c1b-ee56-4786-993f-36682b69025b)

**Comment by raggi** [comment]
thanks @florian-obradovic very informative, I'll definitely look into these details.

I was just able to capture a core part of the transient DNS issue, at some point `socketfilterfw` stops being able to resolve peer programs correctly, this is something I saw Crowstrike reporting affecting them, and this may be one of the items fixed in 15.1.

Here's a log capture from an event when mdnsResponder makes a connection that socketfilterfw decides is bad, after this occurs the firewall blocks all traffic replying to mdnsResponder (restarting mdnsResponder is not sufficient to fix the state):

![image](https://github.com/user-attachments/assets/7b6b7a96-4dc5-4c4e-b727-26c7cc5d3f2d)

In this image, mdnsResponder is pid 422. Attempting to explicitly add mdnsResponder to the list of allowed apps completes without error, but does not add the program to the list, even on a system with SIP disabled.

Forcefully restarting socketfilterfw is also insufficient to unblock mdnsResponder - but a reboot will heal this state for a time.

**Comment by McSim85** [comment]
I have the same issue.
Happy to provide more details.
I tried AppStore Tailscale version and also https://pkgs.tailscale.com/stable/Tailscale-latest-macos.pkg.
Same behaviour. 
When Tailscale connected, it cuts off other internet services.

**Comment by raggi** [comment]
@McSim85 so far there's no permanent mitigation we've been able to verify. If you reboot the machine the issue will be healed for some time. We've now witnessed socketfilterfw even occasionally blocking all incoming traffic to mdnsresponder (both internal macos components), at which point no DNS will work again on the machine until reboot. For some reason this bug triggers more often against Tailscale when it is serving DNS, we're not yet been able to identify why. It appears that the 15.1 beta has a fix that avoids the issue, we're hoping Apple will publish a point release for 15.0 soon. In the meantime the latest unstable has a fix for some extra DNS traffic we identified while diagnosing this bug, and that may help reduce the frequency of triggering the bug, but that is unverified.

**Comment by zmre** [comment]
I believe I'm having these same issues. Symptoms are identical.  Ping works fine, DNS goes to hell.  I'm on 15.0.1.  The problems have become noticeably worse in the last two days since the new version of Tailscale (1.76.1, app store variant) was installed. I'm not certain it's a tailscale issue, but turning off tailscale partially resolves things, but then they degrade further until I have to reboot.  I typically almost never reboot, but currently needing to do it daily.

I use tailscale a lot, but my next move will be to uninstall and then wait to see if that seems to resolve (or lessen) the issue.  I realize this could be an issue in MacOS that isn't the fault of Tailscale, but I'm hoping the community can figure out workarounds.  I've tried `kill -9` on socketfilterfw, mDNSResponder, and mDNSResponderHelper and some things sort of start working, but others don't.  For example, at the moment I can ping and do nslookups from the terminal and my browser is working okay, but Slack, even restarted, can't reach anything. Private relay is off.

I do also have Little Snitch in my environment and have tried to shut it off as well in case it's playing a role here.  When in this state, Tailscale itself is getting Network Map Response Timeout (mapresponse-timeout).

One more bit of info that might trigger something for others: when my computer is in this state, ssh also stops working. I can't git pull, etc. This is the error:

```
ssh: connect to host github.com port 22: Can't assign requested address
fatal: Could not read from remote repository.
```

None of the usual reasons for that apply here.  Maybe I'm out of new connection handles or something?  It doesn't fix itself though.

Happy to try things when I'm next in this state. About to reboot to get back to the good place for another 24 hours or so.  Suggestions welcome.  And again, I realize this probably isn't the fault of Tailscale, though the upgrade to the latest version and the serious uptick in incidences of the problem are a notable coincidence...

**Comment by zmre** [comment]
Still trying to help others by adding info: I've kept tailscale off for the last week and have not re-entered the Bad State.  I'm going to update to 15.1 later today and then turn tailscale back on and see what happens and I'll report back.  Little Snitch has been on this whole time so I don't believe it is triggering the issue.  I still believe it's a MacOS bug of some kind triggered by tailscale and I'm still hunting for workarounds.

**Comment by zmre** [comment]
It's been about 3 days since I updated and have not yet experienced the problem.  Tailscale is running.  Hopeful this means 15.1 resolved the issue.  At a minimum, the frequency has decreased.

**Comment by raggi** [comment]
@cleanbit can you attempt rebuilding your network profiles following a path similar to https://github.com/tailscale/tailscale/issues/13511#issuecomment-2360239552 - where this has been done the issues of the app firewall making spurious decisions has ceased. if that works, please report this back to Apple, as the issue is in their application firewall.

**Comment by cleanbit** [comment]
@raggi yay. I've deleted the VPN profile of Tailscale, restart the app, it installed the profile again, and everything seems to be working.

**Comment by agottardo** [comment]
We'll be closing this as we are not currently aware of further DNS-related issues caused by macOS Sequoia, provided users update to the latest available version, 15.1. If you are experiencing any DNS issue, please file a separate issue attaching a bug report identifier and a macOS Configuration Report.
